packages/react-form-state/src/FormState.ts

```diff
@@ -0,0 +1,351 @@
+import * as React from 'react';
+import isEqual from 'lodash/isEqual';
+import isArray from 'lodash/isArray';
+import memoize from 'lodash/memoize';
+
+import {mapObject} from './utilities';
+import {FieldDescriptors, ClientError, FieldState} from './types';
+import {List, Nested} from './components';
+
+const ERROR_SEPARATOR = ', ';
+
+interface GraphQLError {
+  field: string[] | null;
+  message: string;
+}
+
+export type FormError = GraphQLError | ClientError;
+
+export type FieldStates<Fields> = {
+  [FieldPath in keyof Fields]: FieldState<Fields[FieldPath]>
+};
+
+type MaybeArray<T> = T | T[];
+type MaybePromise<T> = Promise<T> | T;
+
+interface SubmitHandler<Fields> {
+  (fields: FieldStates<Fields>): MaybePromise<FormError[]> | MaybePromise<void>;
+}
+
+export type ValidatorDictionary<Fields> = {
+  [FieldPath in keyof Fields]: MaybeArray<
+    ValidationFunction<Fields[FieldPath], Fields>
+  >
+};
+
+interface ValidationFunction<Value, Fields> {
+  (value: Value, fields: FieldStates<Fields>): any;
+}
+
+export interface FormDetails<Fields> {
+  fields: FieldDescriptors<Fields>;
+  dirty: boolean;
+  valid: boolean;
+  submitting: boolean;
+  errors: ClientError[];
+  reset(): void;
+  submit(): void;
+}
+
+interface Props<Fields> {
+  initialValues: Fields;
+  validators?: Partial<ValidatorDictionary<Fields>>;
+  onSubmit?: SubmitHandler<Fields>;
+  children(form: FormDetails<Fields>): React.ReactNode;
+}
+
+interface State<Fields> {
+  submitting: boolean;
+  errors: ClientError[];
+  dirtyFields: (keyof Fields)[];
+  fields: FieldStates<Fields>;
+}
+
+export default class FormState<
+  Fields extends Object
+> extends React.PureComponent<Props<Fields>, State<Fields>> {
+  static List = List;
+  static Nested = Nested;
+
+  static specialize<FormFields>() {
+    return (FormState as any) as new () => FormState<FormFields>;
+  }
+
+  static getDerivedStateFromProps<T>(newProps: Props<T>, oldState?: State<T>) {
+    const newInitialValues = newProps.initialValues;
+
+    if (oldState == null) {
+      const newState = createFormState(newInitialValues);
+      return newState;
+    }
+
+    const oldInitialValues = initialValuesFromFields(oldState.fields);
+    const shouldReinitialize = !isEqual(oldInitialValues, newInitialValues);
+
+    if (shouldReinitialize) {
+      const newState = createFormState(newInitialValues);
+      return newState;
+    }
+
+    return oldState;
+  }
+
+  state = this.cleanFormState;
+  private mounted = false;
+
+  private getFieldUpdater = memoize(<K extends keyof Fields>(fieldPath: K) => {
+    return this.updateField.bind(this, fieldPath);
+  });
+
+  private getFieldBlurHandler = memoize(
```

Comment:
I find the mix of methods and properties really distracting (also the order is wonky)